src/heretic/main.py

```diff
@@ -49,6 +51,106 @@
 )
 
 
+def get_merged_model(model: Model, settings: Settings):
+    """
+    Returns the model with LoRA adapters merged if applicable.
+    For quantized models, prompts the user for how to proceed.
+    Returns None if user chooses to save LoRA adapter only.
+    """
+    if not isinstance(model.model, PeftModel):
+        # No LoRA adapters, return the model as-is
+        return model.model
+
+    # Prompt for all PEFT models to ensure user is aware of merge implications
+    if settings.quantization in [
+        QuantizationMethod.BNB_4BIT,
+        QuantizationMethod.BNB_8BIT,
+    ]:
+        # Quantized models need special handling - we must reload the base model
+        # in full precision to merge the LoRA adapters
+        print()
+        print(
+            "[yellow]Model was loaded with quantization. Merging requires reloading the base model.[/]"
+        )
+        print(
+            "[red](!) WARNING: CPU Merging requires dequantizing the entire model to System RAM.[/]"
+        )
+        print("[red]    This can lead to SYSTEM FREEZES if you run out of memory.[/]")
+        print(
+            "[yellow]    Rule of thumb: You need approx. 3x the parameter count in GB.[/]"
+        )
+        print(
+            "[yellow]    Example: A 27B model requires ~80GB RAM. A 70B model requires ~200GB RAM.[/]"
+        )
+        print()
+
+        merge_choice = prompt_select(
+            "How do you want to proceed?",
+            choices=[
+                Choice(
+                    title="Merge full model (reload base model on CPU - requires high RAM)",
+                    value="merge",
+                ),
+                Choice(
+                    title="Save LoRA adapter only (can be merged later with llama.cpp or more RAM)",
+                    value="adapter",
+                ),
+            ],
+        )
+
+        if merge_choice is None or merge_choice == "adapter":
+            print("[yellow]Saving LoRA adapter only...[/]")
+            return None
+
+        # User chose to merge - attempt CPU-based merge
+        from transformers import AutoModelForCausalLM
+
+        # Get the adapter state dict before we do anything
+        adapter_state = {}
+        for name, param in model.model.named_parameters():
+            if "lora_" in name:
+                adapter_state[name] = param.data.clone().cpu()
+
+        # Load base model in full precision on CPU to avoid VRAM issues
+        print("* Loading base model on CPU (this may take a while)...")
+        base_model = AutoModelForCausalLM.from_pretrained(
+            settings.model,
+            torch_dtype=torch.bfloat16,
```

Comment:
This should use the `dtype` setting, no? But then some models default to float32 with auto, and some have mixed precision... not sure what the best choice is here.